beacon-chain/blockchain/forkchoice/process_attestation.go

```diff
@@ -93,15 +96,30 @@ func (s *Store) OnAttestation(ctx context.Context, a *ethpb.Attestation) (uint64
 		return 0, err
 	}
 
-	// Use the target state to to validate attestation and calculate the committees.
-	indexedAtt, err := s.verifyAttestation(ctx, baseState, a)
-	if err != nil {
-		return 0, err
-	}
+	// Process aggregated attestation in the queue every `slot/2` duration,
+	// this allows the incoming attestation to aggregate and avoid
+	// process individual attestation.
+	if uint64(time.Now().Unix())%params.BeaconConfig().SecondsPerSlot/2 == 0 {
+		s.attsQueueLock.Lock()
+		defer s.attsQueueLock.Unlock()
+		for root, a := range s.attsQueue {
+			log.WithFields(logrus.Fields{
+				"AggregatedBitfield": fmt.Sprintf("%b", a.AggregationBits),
+				"Root":               fmt.Sprintf("%#x", root),
+			}).Info("Updating latest votes")
 
-	// Update every validator's latest vote.
-	if err := s.updateAttVotes(ctx, indexedAtt, tgt.Root, tgt.Epoch); err != nil {
-		return 0, err
+			// Use the target state to to validate attestation and calculate the committees.
+			indexedAtt, err := s.verifyAttestation(ctx, baseState, a)
+			if err != nil {
+				return 0, err
+			}
+
+			// Update every validator's latest vote.
+			if err := s.updateAttVotes(ctx, indexedAtt, tgt.Root, tgt.Epoch); err != nil {
+				return 0, err
+			}
+			delete(s.attsQueue, root)
```

Comment:
Aren't you receiving multiple attestations with the same data root to aggregate? Why would you delete based off the data root?

Reply:
Because it's processed, example:
say t is the second, and we process aggregated attestations from the queue every 3 secs
```
t10: {rootA: {bitfield:11000000, sig: X}}
t11: {rootA: {bitfield:11100000, sig: Y}}
t12: {rootA: {bitfield:11110000, sig: Z}}
<Process {rootA: {bitfield:11110000, sig: Z}}>
t13: {rootA: {bitfield:00001100, sig: M}}
t14: {rootA: {bitfield:00001111, sig: N}}
t15: {rootB: {bitfield:11000011, sig: X}}
<Process {rootA: {bitfield:00001111, sig: N}}>
<Process {rootB: {bitfield:11000011, sig: X}>
```

Reply:
Oh I see they're already aggregated at this point, thanks!